### ISSUE TITLE
Feature request: Disallow uninitialized properties

### ISSUE THREAD

**Issue by rask** [phpstan/phpstan-strict-rules#88]
With PHP 7.4 typed properties, following is possible:

```php
class Foo
{
    public Bar $bar;
}

$foo = new Foo();
var_dump($foo->bar);
```

Which produces an error.

Would it be useful to have a check that properties may be uninitialized _only_ during `__constructor`, and if the constructor does not set a value it is reported as an error?

**Comment by rask** [comment]
Clarification: when typed as `?type` the property is defaulted to `null`, which is not the same as uninitialized, and which should not trigger the error.

**Comment by ondrejmirtes** [comment]
Hi, this is difficult to get right, because PHP specifically allows the property to be uninitialized until used for the first time. So the property doesn't have to be initialized in the constructor... Any opinion how to tackle this?

**Comment by rask** [comment]
Is the following possible:

1. When loading a class, collect all typed properties which are not nullable or do not contain a default value.
2. When analysing the `__construct()` method, see which of those variables are receiving a value.

Or would this require extensive juggling inside the analyzer code? Of course this could end up deep when we start to factor in calling other class methods that do the property setting, or if some external entity is used to set the value.

**Comment by odan** [comment]
It looks like that the default value of a `?type` property is not `null`, it is still uninitialized.

https://3v4l.org/oanOO

Only by setting `null` directly works.

`public ?Bar $bar = null;`

https://3v4l.org/KqLoA

**Comment by ondrejmirtes** [comment]
See this article: https://phpstan.org/blog/detecting-unused-private-properties-methods-constants
